### PR TITLE
Further hash function improvements

### PR DESCRIPTION
**Pull request by ChrisJefferson**
Some general cleanups and improvements. Some also (shockingly) bad documentation, but I'm thinking something is better than nothing.

### CODE REVIEW

File: src/hashfunctions.c [hunk 1]
```diff
@@ -139,13 +139,16 @@ Obj DATA_HASH_FUNC_FOR_INT(Obj self, Obj i)
         return HashValueToObjInt(DataHashFuncForInt(i));
 }
 
-Int BasicRecursiveHash(Obj obj);
+static inline Int BasicRecursiveHash(Obj obj);
+
+// This is just a random number which fits in a 32-bit UInt.
+// It is used the base for hashing of lists
+#define LIST_BASE_HASH 2195952830L
```
Comment (fingolfin) [inline src/hashfunctions.c]: Nitpick, but: I tend to prefer enums over macros; this way the compiler sees the name.

File: src/hashfunctions.c [hunk 2]
```diff
@@ -208,34 +218,60 @@ Int BasicPrimitiveHash(Obj obj)
         return HashFuncForTrans(obj);
     }
 
+     if (IS_LIST(obj)) {
+        return BasicRecursiveHashForList(obj);
+    }
+    else if (IS_PREC_REP(obj)) {
+        return BasicRecursiveHashForPRec(obj);
```
Comment (fingolfin) [inline src/hashfunctions.c]: Why not put this in the switch? There are just two TNUMs for this.

File: src/hashfunctions.c [hunk 3]
```diff
@@ -208,34 +218,60 @@ Int BasicPrimitiveHash(Obj obj)
         return HashFuncForTrans(obj);
     }
 
+     if (IS_LIST(obj)) {
```
Comment (fingolfin) [inline src/hashfunctions.c]: Bad indentation

File: src/hashfunctions.c [hunk 4]
```diff
@@ -261,6 +295,17 @@ static Int InitKernel(void)
 static Int InitLibrary(void)
 {
     InitGVarFuncsFromTable(GVarFuncs);
+
+    // We  Reuse the 1 argument function for 1 argument, and also for
```
Comment (fingolfin) [inline src/hashfunctions.c]: "We  Reuse" ???

File: src/hashfunctions.c [hunk 5]
```diff
@@ -261,6 +295,17 @@ static Int InitKernel(void)
...
+    // handling a list of arguments
+    Obj gvar = NewFunctionC("DATA_HASH_FUNC_RECURSIVE", -1L, "arg",
```
Comment (fingolfin) [inline src/hashfunctions.c]: Why -1L and not just -1?

Comment (ChrisJefferson) [inline src/hashfunctions.c]: Just blindly copied it from some other call of `NewFunctionC`, will have a look if there is any reason this is actually required.

**Comment by ChrisJefferson** [comment]
Fixed another comment on testing, the only thing I haven't done is move things around, mainly because I'm not sure where they should go, and this is fairly awful (at least it should be good enough for someone else to know what I've done :) )

**Pull request merged**